src/Raven.Client/Http/RequestExecutor.cs

```diff
@@ -44,58 +48,75 @@ public class AggresiveCacheOptions
 
         public readonly HttpCache Cache = new HttpCache();
 
-        private Topology _topology;
+        private int _hasUpdatedTopologyAtLeastOnce = 0;
+
+        public IReadOnlyList<ServerNode> TopologyNodes => _nodeSelector.Topology.Nodes;
+
         private readonly Timer _updateTopologyTimer;
 
         private Timer _updateCurrentTokenTimer;
-        private readonly Timer _updateFailingNodesStatus;
 
-        public string Url => _topology.LeaderNode.Url;
+        private readonly NodeSelector _nodeSelector;
+
+        //note: the condition for non empty nodes is precaution, should never happen..
+        public string Url => _nodeSelector.CurrentNode?.Url;
+        
+        public event Action TopologyUpdate;
+
+        public bool HasUpdatedTopologyOnce { get; private set; }
 
         private RequestExecutor(string url, string databaseName, string apiKey, bool requiresTopologyUpdates)
         {
+            _initialUrl = url;
             _apiKey = apiKey;
-            _topology = new Topology
+            _requiresTopologyUpdates = requiresTopologyUpdates;
+            _nodeSelector = new NodeSelector(new Topology
             {
-                LeaderNode = new ServerNode
+                Nodes = new List<ServerNode>
                 {
-                    Url = url,
-                    Database = databaseName,
+                    new ServerNode
+                    {
+                        Url = url,
+                        Database = databaseName,
+                        ClusterToken = "N/A" //initial topology before an update, so cluster token is not applicable here...
+                    }
                 },
-                ReadBehavior = ReadBehavior.LeaderOnly,
-                WriteBehavior = WriteBehavior.LeaderOnly,
                 Etag = int.MinValue,
                 SLA = new TopologySla
                 {
                     RequestTimeThresholdInMilliseconds = 100
                 }
-            };
+            });
 
             ContextPool = new JsonContextPool();
 
             if (requiresTopologyUpdates == false)
                 return;
 
             _updateTopologyTimer = new Timer(UpdateTopologyCallback, null, 0, Timeout.Infinite);
-            _updateFailingNodesStatus = new Timer(UpdateFailingNodesStatusCallback, null, TimeSpan.FromMinutes(1),
-                TimeSpan.FromMinutes(1));
+
+            //this would be handled by heartbeats between cluster nodes (that are not yet implemented)
+            //_updateFailingNodesStatus = new Timer(UpdateFailingNodesStatusCallback, null, TimeSpan.FromMinutes(1),
```

Comment:
This is not part of the client. The client merely update its topology every minutes, nothing more.